File: cuttingx/cuttingx/report/cutting_completion_report/cutting_completion_report.js
// Copyright (c) 2025, Cognitonx Logic India Private limited and contributors
// For license information, please see license.txt

frappe.query_reports["Cutting completion Report"] = {
    hasRole(role) {
        if (frappe.user?.has_role) return frappe.user.has_role(role);
        const roles =
            frappe.boot?.user_info?.[frappe.session.user]?.roles || frappe.user_roles || [];
        return Array.isArray(roles) && roles.includes(role);
    },

    formatter(value, row, column, data, default_formatter) {
        const html = default_formatter(value, row, column, data, default_formatter);
        if (!data) return html;

        const fieldname = (column.fieldname || "").toLowerCase();
        const isStatus = fieldname === "status";
        const isFolding = fieldname === "folding";
        const isEndBit = fieldname === "end_bit";

        // Handle Folding and End Bit
        if (isFolding || isEndBit) {
            const docname = data.can_cut_name;
            if (!docname) return html;  // ← Will be empty if no Can Cut exists

            const safeValue = frappe.utils.escape_html(value || "");
            return `
                <textarea class="report-editable-field"
                          data-docname="${docname}"
                          data-doctype="Can Cut"
                          data-fieldname="${fieldname}"
                          rows="1"
                          style="width:100%; padding:4px; resize:vertical;">${safeValue}</textarea>
            `;
        }

        // Handle Status Dropdown
        if (isStatus) {
            const docname = data.ocn;
            const currentValue = value || "Pending for Approval";
            const isFactoryManager = this.hasRole("Factory Manager");

            const opts = ["Pending for Approval"];
            if (isFactoryManager) opts.push("Approved");

            if (data.is_first_row) {
                const options = opts.map(opt =>
                    `<option value="${opt}" ${opt === currentValue ? "selected" : ""}>${opt}</option>`
                ).join("");

                return `
                    <select class="report-status-select"
                            data-docname="${docname}"
                            data-doctype="Sales Order"
                            data-fieldname="custom_consumption_status"
                            style="width:100%; padding:4px; border-radius:4px;">
                        ${options}
                    </select>
                `;
            } else {
                return `<span>${frappe.utils.escape_html(currentValue)}</span>`;
            }
        }

        return html;
    },

    onload(report) {
        const $wrap = report.page.wrapper;

        setTimeout(() => {
            const columns = report.get_columns() || [];
            columns.forEach(c => {
                if (["status", "folding", "end_bit"].includes((c.fieldname || "").toLowerCase())) {
                    c.editable = 1;
                }
            });
        }, 500);

        const save = frappe.utils.debounce(function (e) {
            const $el = $(e.currentTarget);
            const docname = $el.data("docname");
            const doctype = $el.data("doctype");
            const fieldname = $el.data("fieldname");
            const value = $el.val();

            // Handle Approval via Custom API
            if (fieldname === "custom_consumption_status" && value === "Approved") {
                const isAllowed = this.hasRole("Factory Manager");
                if (!isAllowed) {
                    frappe.msgprint(__("Only Factory Manager can set status to 'Approved'"));
                    $el.val($el.data("old-value"));
                    return;
                }

                // Get all rows for this OCN
                const ocn = docname;
                const relatedRows = report.data.filter(row => row.ocn === ocn);

                // Define required fields
                const requiredFields = [
                    { key: "fabric_ordered", label: "Fabric Ordered" },
                    { key: "fabric_issued", label: "Fabric Issued" },
                    { key: "folding", label: "Folding" },
                    { key: "end_bit", label: "End Bit" },
                    { key: "file_consumption", label: "File Consumption" },
                    { key: "actual_consumption", label: "Actual Consumption" }
                ];

                const missingFields = [];

                relatedRows.forEach(row => {
                    requiredFields.forEach(field => {
                        const val = row[field.key];
                        if (!val || String(val).trim() === "") {
                            if (!missingFields.includes(field.label)) {
                                missingFields.push(field.label);
                            }
                        }
                    });
                });

                if (missingFields.length > 0) {
                    const message = `Cutting flow not completed. Cannot approve.<br><br>Missing: <b>${missingFields.join(", ")}</b>`;
                    frappe.msgprint({
                        title: __('Approval Blocked'),
                        indicator: 'red',
                        message: __(message)
                    });
                    $el.val($el.data("old-value"));
                    return;
                }

                // Show confirmation dialog
                frappe.confirm(
                    __("Approve this Sales Order? This will finalize the cutting status."),
                    () => {
                        // User clicked Yes
                        $el.css("opacity", 0.6);
                        frappe.call({
                            method: "erpnext_trackerx_customization.api.approve_consumption_status.approve_consumption_status",
                            args: {
                                sales_order: docname
                            },
                            callback: (r) => {
                                if (!r.exc) {
                                    frappe.show_alert({ message: __("Approved!"), indicator: "green" });
                                    $el.data("old-value", "Approved");
                                    report.refresh();  // Reload data to show approved_by/on
                                } else {
                                    // Show error from server
                                    $el.val($el.data("old-value"));
                                }
                            },
                            always: () => {
                                $el.css("opacity", 1);
                            }
                        });
                    },
                    () => {
                        // User clicked No / Cancel
                        $el.val($el.data("old-value"));
                    }
                );

                return;  // Prevent default save
            }

            // Default save for other fields (e.g., Folding, End Bit)
            $el.css("opacity", 0.6);
            frappe.call({
                method: "frappe.client.set_value",
                args: { doctype, name: docname, fieldname, value },
                callback(r) {
                    if (!r.exc) {
                        frappe.show_alert({ message: __("Saved"), indicator: "green" });
                        $el.data("old-value", value);
                    } else {
                        frappe.msgprint(__("Save failed"));
                    }
                },
                always() {
                    $el.css("opacity", 1);
                }
            });
        }.bind(this), 600);  // Bind `this` for role check

        // Track original value
        $wrap.on("focus", ".report-editable-field, .report-status-select", function () {
            $(this).data("old-value", $(this).val());
        });

        // Save on blur (text areas) and change (dropdowns)
        $wrap.on("blur", ".report-editable-field", save);
        $wrap.on("change", ".report-status-select", save);
    }
};

File: cuttingx/cuttingx/report/cutting_completion_report/cutting_completion_report.py
# Copyright (c) 2025, Cognitonx Logic India Private limited and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {
            "label": _("OCN"),
            "fieldname": "ocn",
            "fieldtype": "Link",
            "options": "Sales Order",
            "width": 120
        },
        {
            "label": _("Style"),
            "fieldname": "style",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Colour"),
            "fieldname": "colour",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Order Qty"),
            "fieldname": "order_qty",
            "fieldtype": "Int",
            "width": 100
        },
        {
            "label": _("Fabric Ordered"),
            "fieldname": "fabric_ordered",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "label": _("Fabric Issued"),
            "fieldname": "fabric_issued",
            "fieldtype": "Float",
            "width": 140
        },
        {
            "label": _("Folding"),
            "fieldname": "folding",
            "fieldtype": "Small Text",
            "width": 120
        },
        {
            "label": _("End Bit"),
            "fieldname": "end_bit",
            "fieldtype": "Small Text",
            "width": 120
        },
        {
            "label": _("File Consumption"),
            "fieldname": "file_consumption",
            "fieldtype": "Float",
            "width": 140
        },
        {
            "label": _("Actual Consumption"),
            "fieldname": "actual_consumption",
            "fieldtype": "Float",
            "width": 160
        },
        {
            "label": _("Can Cut Qty"),
            "fieldname": "can_cut_qty",
            "fieldtype": "Float",
            "width": 100
        },
        {
            "label": _("Cut Qty Actual"),
            "fieldname": "cut_qty_actual",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "label": _("Difference"),
            "fieldname": "difference",
            "fieldtype": "Float",
            "width": 100
        },
        {
            "label": _("Status"),
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Approved By"),
            "fieldname": "custom_approved_by",
            "fieldtype": "Link",
            "options": "User",
            "width": 140
        },
        {
            "label": _("Approved On"),
            "fieldname": "custom_approved_on",
            "fieldtype": "Datetime",
            "width": 160
        }        
    ]

def get_data(filters):
    conditions = ""
    if filters.get("from_date"):
        conditions += " AND so.delivery_date >= %(from_date)s"
    if filters.get("to_date"):
        conditions += " AND so.delivery_date <= %(to_date)s"

    query = """
        SELECT 
            *,
            CASE WHEN rn = 1 THEN 1 ELSE 0 END AS is_first_row
        FROM (
            SELECT
                so.name AS ocn,
                item.custom_style_master AS style,
                sod.custom_color AS colour,
                SUM(sod.custom_order_qty) AS order_qty,  -- Aggregated
                so.delivery_date,

                cc.fabric_ordered,
                cc.fabric_issued,
                cc.folding,
                cc.end_bit,
                cc.file_consumption,
                cc.actual_consumption,
                cc.name AS can_cut_name,

                CASE 
                    WHEN cc.actual_consumption > 0 THEN (cc.fabric_issued / cc.actual_consumption)
                    ELSE 0
                END AS can_cut_qty,

                COALESCE((
                    SELECT SUM(cci.confirmed_quantity)
                    FROM `tabCut Confirmation Item` cci
                    INNER JOIN `tabCut Confirmation` con ON con.name = cci.parent
                    INNER JOIN `tabCut Docket` cd ON cd.name = con.cut_po_number
                    WHERE cci.sales_order = so.name
                    AND cd.color = sod.custom_color
                    AND cci.docstatus = 1
                ), 0) AS cut_qty_actual,

                (COALESCE((
                    SELECT SUM(cci.confirmed_quantity)
                    FROM `tabCut Confirmation Item` cci
                    INNER JOIN `tabCut Confirmation` con ON con.name = cci.parent
                    INNER JOIN `tabCut Docket` cd ON cd.name = con.cut_po_number
                    WHERE cci.sales_order = so.name
                    AND cd.color = sod.custom_color
                    AND cci.docstatus = 1
                ), 0) - SUM(sod.custom_order_qty)) AS difference,

                COALESCE(so.custom_consumption_status, 'Pending for Approval') AS status,

                so.custom_approved_by,
                so.custom_approved_on,                

                ROW_NUMBER() OVER (PARTITION BY so.name ORDER BY sod.custom_color) AS rn

            FROM `tabSales Order` so
            INNER JOIN `tabSales Order Item` sod ON sod.parent = so.name
            INNER JOIN `tabItem` item ON item.name = sod.item_code
            LEFT JOIN `tabCan Cut` cc 
                ON cc.sales_order = so.name 
                AND cc.colour = sod.custom_color
            WHERE so.docstatus = 1
            {conditions}

            GROUP BY so.name, sod.custom_color, item.custom_style_master, cc.name
            ORDER BY so.delivery_date, so.name, sod.custom_color
        ) sub_query
        ORDER BY delivery_date, ocn, rn
    """.format(conditions=conditions)

    data = frappe.db.sql(query, filters, as_dict=1)
    return data